### DISCUSSION TITLE
Nullable analysis for a collection of nullable reference types

### DISCUSSION THREAD

**Discussion by Vannevelj** [dotnet/roslyn#64246] [General]
We can use `[NotNull]` to create our own checks that affect nullability flow. I have a file with a bunch of these, some of which operate on a single value (`Guard.AgainstNull(string param)`) while others verify multiple: `Guard.AgainstNull(List<string> ids)`

As I'm adding NRT to the project I can use `[NotNull]` for the former but it wouldn't work for the `List<T>` as I cannot target the individual values specifically. A quick example for that is here:

```cs
using System.Diagnostics.CodeAnalysis;

List<string?> values = new();
Check(values);

string trimmed = values.First().Trim(); // Dereference of a possible null reference

void Check([NotNull] IEnumerable<string?> values)
{
    if (values.Any(x => x == null))
    {
        throw new ArgumentNullException(nameof(values));
    }
}
```

Is there a workaround for this yet? Passing a list of IDs is a very common pattern for us and I'd like to avoid having to copy them into a new collection just to make the flow checker happy.

**Comment by RikkiGibson** [comment]
A variable's type arguments can't change based on flow analysis. You'll need to do the sufficient null checks, then do an "unsafe" conversion of the input to the result type. [SharpLab](https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzggPgYhgVwDZ4TB4CmABCTEaQLABQ9AAgEwCMTADGY6wCwDc9JgGZuzMgGEyAb3pl53UY15kAsgAoAMlDQAeHhwD8APm6sOKAJQy5CuwDcIAJzIBjABYlXAaxIATAGUEJ1gAcxQyAF5JTx91AytBOjs7BIBtDgBdADoAFQB7IJCYUPVLfjIAekqyAHdnGDDbFLdY30DgsJQMnIKisLKK6rJ872ayAF9x8e09XNMJNt15rR0EZZMyPDXLcdlklrIAM3ynEggPMnVHFygEEjAyWC2d8bt9w5SoI6u7h6eIrgCNZpGQEO4nPlamQYCRoQBBJyhHBgSgIABy+DwAFEAB6uEgABwQUHyMEGkzeCimBxajAA7C80ABCJJ2GkTIA=).

```cs
#nullable enable

using System;
using System.Collections.Generic;

public class C {
    public void M(List<string?> strings) {
        var checkedStrings = Check(strings);
        strings[0].ToString(); // warning
        checkedStrings[0].ToString(); // ok
    }
    
    List<T> Check<T>(List<T?> list)
    {
        foreach (var item in list)
        {
            if (item is null) { throw new ArgumentNullException(); }
        }
        return list!;
    }
}
```

I'm not actually sure how to make it work on an arbitrary `IEnumerable<T>` while preserving the type information in the output. I thought that some set of type parameters and constraints would make it possible, but couldn't figure it out at a glance.

**Reply by Vannevelj** [reply]
> A variable's type arguments can't change based on flow analysis

Is that because it just hasn't been pulled through that far yet, or is there something innate that prohibits this? I don't think this target specifier exists but if we could do something along these lines that would be ideal:

```cs
void Check([typearg1: NotNull] IEnumerable<string?> values)
```

I can work with your suggestion since it avoids the copy although it does feel like a step back for our developers. A slight riff on that appeals to me a bit more:

```cs
void AgainstNull<T>(IEnumerable<T?> values, out IEnumerable<T> validated)
{
    if (values.Any(x => x == null))
    {
        throw new ArgumentNullException(nameof(values));
    }

    validated = values!;
}
```

**Reply by RikkiGibson** [reply]
> Is that because it just hasn't been pulled through that far yet, or is there something innate that prohibits this?

No work is planned to permit nested nullable conversions based on flow state in general. There's a chance that work related to #37468 could improve the experience here, but it's not clear if we will be able to prioritize that work. It's not to say that the issue is unimportant. We just have a lot of issues that compete for a limited amount of developer time. Certainly people's feedback does help us figure out what to spend time on. If you think some specialized nullable tracking for collection values could help here, feel free to weigh in on that issue.

I also don't think attributes on type arguments is forthcoming, though I've also had cases where I've wanted them unrelated to nullable. There's not a position in metadata to store them, so someone would have to design an encoding of some kind, and ensure that the encoding would be handled by various metadata reading and writing tools in the ecosystem in a reasonably elegant way.

The limitation around type arguments is more innate because they can be used to describe the behavior of methods on the type in addition to just types of fields. Imagine a user creates an instance of `Deserializer<MyType?>`. The user is telling us they think the values produced by the deserializer might be null. It wouldn't make sense to convert this to `Deserializer<MyType>` even though the instance does not store any values of type `MyType?`. So we'd have to come up with a scheme for deciding when the conversion is safe depending on some kind of annotation on the type parameter. That's why we've just landed on the experience of asking the user to suppress the conversion warning if they think the conversion is the right thing to do.
